### PR TITLE
ContiguousCollection protocol revisited

### PR DESCRIPTION
**Pull request by airspeedswift**
A re-pitch of the protocols dropped from SE-0237, with motivating example.

### CODE REVIEW

File: proposals/NNNN-contiguous-collection.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+buffer.
+
+```swift
+/// A collection that supports access to its underlying contiguous storage.
+public protocol ContiguousCollection: Collection
+where SubSequence: ContiguousCollection {
+  /// Calls a closure with a pointer to the array's contiguous storage.
+  func withUnsafeBufferPointer<R>(
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+    ) rethrows -> R
```
Comment (stephentyrone) [inline proposals/NNNN-contiguous-collection.md]: `)` should de-indent here, no? Also below.

**Comment by airspeedswift** [comment]
Oops in fact that was a typo, I meant `vDSP_vaddi`. The fact that they both take a pointer is super cursed. Switched to the `Float` version.

**Comment by stephentyrone** [comment]
Fortran. Boy, I don’t know.

File: proposals/NNNN-contiguous-collection.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  mutating func withUnsafeMutableBufferPointer<R>(
+    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R
+}
+```
+
+Conformances will be added for the following types:
+
+- `Array`, `ArraySlice` and `ContiguousArray` will conform to `MutableContiguousCollection`
+- `UnsafeBufferPointer` will conform to ContiguousCollection
```
Comment (moiseev) [inline proposals/NNNN-contiguous-collection.md]:
```suggestion
- `UnsafeBufferPointer` will conform to `ContiguousCollection`
```

File: proposals/NNNN-contiguous-collection.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
+# Introduce `{Mutable}ContiguousCollection` protocol
+
+* Proposal: [SE-NNNN](NNNN-contiguous-collection.md)
+* Authors: [Ben Cohen](https://github.com/airspeedswift)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
```
Comment (tkremenek) [inline proposals/NNNN-contiguous-collection.md]: Is there an implementation?

**Pull request merged**